module/zfs/vdev_disk.c

```diff
@@ -83,16 +84,31 @@ vdev_bdev_mode(int smode)
 #endif /* HAVE_OPEN_BDEV_EXCLUSIVE */
 
 static uint64_t
-bdev_capacity(struct block_device *bdev)
+bdev_capacity(struct block_device *bdev, boolean_t wholedisk)
 {
 	struct hd_struct *part = bdev->bd_part;
-
-	/* The partition capacity referenced by the block device */
-	if (part)
-		return (part->nr_sects << 9);
-
-	/* Otherwise assume the full device capacity */
-	return (get_capacity(bdev->bd_disk) << 9);
+	if (part) {
+		if (wholedisk) {
+			/* Total number of sectors on the device */
+			uint64_t sectors = get_capacity(bdev->bd_disk);
+			/* Sectors used by the EFI partition and partition alignment */
+			uint64_t used = EFI_MIN_RESV_SIZE +
+			    NEW_START_BLOCK +
+			    PARTITION_END_ALIGNMENT;
+			if (sectors <= used)
+				return (0);
+			/* Sectors available to the vdev */
+			uint64_t available = sectors - used;
+			/* Bytes available to the vdev */
+			return (available << 9);
+		} else {
+			/* The partition capacity referenced by the block device */
+			return (part->nr_sects << 9);
```

Comment:
nit: while we're here let's use `SECTOR_BITS` instead of `9`.  It's defined by the kernel and will always be 9.

Reply:
Sounds good